File: player_feature/src/main/java/com/parabola/player_feature/AudioEffectsInteractorImpl.java
package com.parabola.player_feature;

import android.media.audiofx.BassBoost;
import android.media.audiofx.Equalizer;
import android.media.audiofx.Virtualizer;

import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.audio.AudioListener;
import com.parabola.domain.interactor.player.AudioEffectsInteractor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

public class AudioEffectsInteractorImpl implements AudioEffectsInteractor {
    private static final String LOG_TAG = AudioEffectsInteractorImpl.class.getSimpleName();

    private final BehaviorSubject<Boolean> playbackSpeedEnabledUpdates;
    private final BehaviorSubject<Float> savedPlaybackSpeed;
    private final BehaviorSubject<Boolean> playbackPitchEnabledUpdates;
    private final BehaviorSubject<Float> savedPlaybackPitch;

    private final BehaviorSubject<Short> savedVirtualizerLevelUpdates;
    private final BehaviorSubject<Short> savedBassBoostLevelUpdates;

    private final SimpleExoPlayer exoPlayer;
    private final PlayerSettingSaver settings;


    //FXs
    private Equalizer equalizer;
    private BassBoost bassBoost;
    private Virtualizer virtualizer;

    private static final float DEFAULT_PLAYBACK_SPEED = 1.0f;
    private static final float DEFAULT_PLAYBACK_PITCH = 1.0f;

    AudioEffectsInteractorImpl(SimpleExoPlayer exoPlayer, PlayerSettingSaver settings) {
        this.exoPlayer = exoPlayer;
        this.settings = settings;

        float speed = settings.getSavedPlaybackSpeedEnabled() ? settings.getSavedPlaybackSpeed() : DEFAULT_PLAYBACK_SPEED;
        float pitch = settings.getSavedPlaybackPitchEnabled() ? settings.getSavedPlaybackPitch() : DEFAULT_PLAYBACK_PITCH;

        PlaybackParameters playbackParameters = new PlaybackParameters(speed, pitch);
        this.exoPlayer.setPlaybackParameters(playbackParameters);

        savedPlaybackSpeed = BehaviorSubject.createDefault(settings.getSavedPlaybackSpeed());
        savedPlaybackPitch = BehaviorSubject.createDefault(settings.getSavedPlaybackPitch());

        playbackSpeedEnabledUpdates = BehaviorSubject.createDefault(settings.getSavedPlaybackSpeedEnabled());
        playbackPitchEnabledUpdates = BehaviorSubject.createDefault(settings.getSavedPlaybackPitchEnabled());

        savedVirtualizerLevelUpdates = BehaviorSubject.createDefault(settings.getSavedVirtualizerStrength());
        savedBassBoostLevelUpdates = BehaviorSubject.createDefault(settings.getSavedBassBoostStrength());

        this.exoPlayer.addListener(new Player.EventListener() {
            @Override
            public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {
                if (savedPlaybackSpeed.getValue() != playbackParameters.speed && settings.getSavedPlaybackSpeedEnabled()) {
                    savedPlaybackSpeed.onNext(playbackParameters.speed);
                }
                if (savedPlaybackPitch.getValue() != playbackParameters.pitch && settings.getSavedPlaybackPitchEnabled()) {
                    savedPlaybackPitch.onNext(playbackParameters.pitch);
                }
            }
        });

        this.exoPlayer.addAudioListener(new AudioListener() {
            @Override
            public void onAudioSessionId(int audioSessionId) {
                try {
                    bassBoost = new BassBoost(0, audioSessionId);
                } catch (RuntimeException ignored) {
                }
                if (bassBoost != null) {
                    bassBoost.setEnabled(settings.getSavedBassBoostEnabled());
                    bassBoost.setStrength(settings.getSavedBassBoostStrength());
                }

                try {
                    virtualizer = new Virtualizer(0, audioSessionId);
                } catch (RuntimeException ignored) {
                }
                if (virtualizer != null) {
                    virtualizer.setEnabled(settings.getSavedVirtualizerEnabled());
                    virtualizer.setStrength(settings.getSavedVirtualizerStrength());
                }

                try {
                    equalizer = new Equalizer(0, audioSessionId);
                } catch (RuntimeException ignored) {
                }
                if (equalizer != null) {
                    equalizer.setEnabled(settings.getSavedIsEqEnabled());
                    for (short i = 0; i < equalizer.getNumberOfBands(); i++) {
                        equalizer.setBandLevel(i, (short) (settings.getSavedBandLevel(i) * 100));
                    }
                }
            }
        });
    }


    //    SPEED
    @Override
    public Observable<Boolean> observeIsPlaybackSpeedEnabled() {
        return playbackSpeedEnabledUpdates;
    }

    @Override
    public void setPlaybackSpeedEnabled(boolean enable) {
        settings.setSavedPlaybackSpeedEnabled(enable);
        playbackSpeedEnabledUpdates.onNext(enable);

        float speed = enable ? settings.getSavedPlaybackSpeed() : DEFAULT_PLAYBACK_SPEED;

        PlaybackParameters playbackParameters = new PlaybackParameters(
                speed, exoPlayer.getPlaybackParameters().pitch, exoPlayer.getPlaybackParameters().skipSilence);
        this.exoPlayer.setPlaybackParameters(playbackParameters);
    }

    @Override
    public float getSavedPlaybackSpeed() {
        return savedPlaybackSpeed.getValue();
    }

    @Override
    public Observable<Float> observePlaybackSpeed() {
        return savedPlaybackSpeed;
    }

    @Override
    public void setSavedPlaybackSpeed(float speed) {
        settings.setPlaybackSpeed(speed);

        if (settings.getSavedPlaybackSpeedEnabled()) {
            PlaybackParameters playbackParameters = new PlaybackParameters(
                    speed, exoPlayer.getPlaybackParameters().pitch, exoPlayer.getPlaybackParameters().skipSilence);

            exoPlayer.setPlaybackParameters(playbackParameters);
        }
    }


    //    PITCH
    @Override
    public Observable<Boolean> observeIsPlaybackPitchEnabled() {
        return playbackPitchEnabledUpdates;
    }

    @Override
    public void setPlaybackPitchEnabled(boolean enabled) {
        settings.setSavedPlaybackPitchEnabled(enabled);
        playbackPitchEnabledUpdates.onNext(enabled);

        float pitch = enabled ? settings.getSavedPlaybackPitch() : DEFAULT_PLAYBACK_PITCH;

        PlaybackParameters playbackParameters = new PlaybackParameters(
                exoPlayer.getPlaybackParameters().speed, pitch, exoPlayer.getPlaybackParameters().skipSilence);
        this.exoPlayer.setPlaybackParameters(playbackParameters);
    }

    @Override
    public float getSavedPlaybackPitch() {
        return savedPlaybackPitch.getValue();
    }

    @Override
    public Observable<Float> observePlaybackPitch() {
        return savedPlaybackPitch;
    }

    @Override
    public void setSavedPlaybackPitch(float pitch) {
        settings.setPlaybackPitch(pitch);

        if (settings.getSavedPlaybackPitchEnabled()) {
            PlaybackParameters playbackParameters = new PlaybackParameters(
                    exoPlayer.getPlaybackParameters().speed, pitch, exoPlayer.getPlaybackParameters().skipSilence);

            exoPlayer.setPlaybackParameters(playbackParameters);
        }
    }


    //    BASS BOOST
    @Override
    public boolean isBassBoostAvailable() {
        return bassBoost != null && bassBoost.getStrengthSupported();
    }

    @Override
    public boolean isBassBoostEnabled() {
        return bassBoost != null && bassBoost.getEnabled();
    }

    @Override
    public void setBassBoostEnable(boolean enable) {
        if (bassBoost != null)
            bassBoost.setEnabled(enable);

        settings.setBassBoostEnabled(enable);
    }

    @Override
    public Observable<Short> observeBassBoostLevel() {
        return savedBassBoostLevelUpdates;
    }

    @Override
    public void setBassBoostLevel(short strength) {
        if (bassBoost != null)
            bassBoost.setStrength(strength);

        settings.setBassBoostStrength(strength);
        savedBassBoostLevelUpdates.onNext(strength);
    }


    //    VIRTUALIZER
    @Override
    public boolean isVirtualizerAvailable() {
        return virtualizer != null && virtualizer.getStrengthSupported();
    }

    @Override
    public boolean isVirtualizerEnabled() {
        return virtualizer != null && virtualizer.getEnabled();
    }

    @Override
    public void setVirtualizerEnable(boolean enable) {
        if (virtualizer != null)
            virtualizer.setEnabled(enable);

        settings.setVirtualizerEnabled(enable);
    }

    @Override
    public Observable<Short> observeVirtualizerLevel() {
        return savedVirtualizerLevelUpdates;
    }

    @Override
    public void setVirtualizerLevel(short strength) {
        if (virtualizer != null)
            virtualizer.setStrength(strength);

        settings.setVirtualizerStrength(strength);
        savedVirtualizerLevelUpdates.onNext(strength);
    }

    //    EQ
    @Override
    public void setEqEnable(boolean enable) {
        if (equalizer != null)
            equalizer.setEnabled(enable);

        settings.setEqEnabled(enable);
    }

    @Override
    public boolean isEqEnabled() {
        return equalizer != null && equalizer.getEnabled();
    }

    @Override
    public short getMaxEqBandLevel() {
        return equalizer != null ? (short) ((equalizer.getBandLevelRange()[1]) / 100) : 0;
    }

    @Override
    public short getMinEqBandLevel() {
        return equalizer != null ? (short) (equalizer.getBandLevelRange()[0] / 100) : 0;
    }

    @Override
    public void setBandLevel(int bandId, short bandLevel) {
        if (equalizer != null)
            equalizer.setBandLevel((short) bandId, (short) (bandLevel * 100));

        settings.setBandLevel((short) bandId, bandLevel);
    }

    @Override
    public List<EqBand> getBands() {
        if (equalizer == null)
            return Collections.emptyList();

        List<EqBand> bands = new ArrayList<>(equalizer.getNumberOfBands());

        for (short i = 0; i < equalizer.getNumberOfBands(); i++) {
            EqBand band = new EqBand();
            band.id = i;
            band.frequency = equalizer.getCenterFreq(i) / 1000;
            band.currentLevel = (short) (equalizer.getBandLevel(i) / 100);
            bands.add(band);
        }

        return bands;
    }
}

File: player_feature/src/main/java/com/parabola/player_feature/AudioEffectsInteractorImpl.kt
package com.parabola.player_feature

import android.media.audiofx.BassBoost
import android.media.audiofx.Virtualizer
import com.google.android.exoplayer2.PlaybackParameters
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.audio.AudioListener
import com.parabola.domain.interactor.player.AudioEffectsInteractor
import com.parabola.domain.interactor.player.AudioEffectsInteractor.EqBand
import com.parabola.player_feature.effect.Effect
import com.parabola.player_feature.effect.EqualizerEffectImpl
import com.parabola.player_feature.effect.EqualizerEffectUnavailable
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject


private const val DEFAULT_PLAYBACK_SPEED = 1.0f
private const val DEFAULT_PLAYBACK_PITCH = 1.0f

internal class AudioEffectsInteractorImpl(
    private val exoPlayer: SimpleExoPlayer,
    private val settings: PlayerSettingSaver,
) : AudioEffectsInteractor {

    private val playbackSpeedEnabledUpdates: BehaviorSubject<Boolean>
    private val savedPlaybackSpeed: BehaviorSubject<Float>
    private val playbackPitchEnabledUpdates: BehaviorSubject<Boolean>
    private val savedPlaybackPitch: BehaviorSubject<Float>
    private val savedVirtualizerLevelUpdates: BehaviorSubject<Short>
    private val savedBassBoostLevelUpdates: BehaviorSubject<Short>


    // TODO [23.06.2022] убрать nullable у переменных, узнать как это выделяется в отдельную абстракцию
    //FXs
    private lateinit var equalizer: Effect.EqualizerEffect
    private var bassBoost: BassBoost? = null
    private var virtualizer: Virtualizer? = null


    init {
        val speed =
            if (settings.isPlaybackSpeedEnabled) settings.playbackSpeed
            else DEFAULT_PLAYBACK_SPEED
        val pitch =
            if (settings.isPlaybackPitchEnabled) settings.playbackPitch
            else DEFAULT_PLAYBACK_PITCH

        exoPlayer.setPlaybackParameters(PlaybackParameters(speed, pitch))

        savedPlaybackSpeed = BehaviorSubject.createDefault(settings.playbackSpeed)
        savedPlaybackPitch = BehaviorSubject.createDefault(settings.playbackPitch)

        playbackSpeedEnabledUpdates =
            BehaviorSubject.createDefault(settings.isPlaybackSpeedEnabled)
        playbackPitchEnabledUpdates =
            BehaviorSubject.createDefault(settings.isPlaybackPitchEnabled)

        savedVirtualizerLevelUpdates =
            BehaviorSubject.createDefault(settings.virtualizerStrength)
        savedBassBoostLevelUpdates = BehaviorSubject.createDefault(settings.bassBoostStrength)

        exoPlayer.addListener(object : Player.EventListener {
            override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {
                if (savedPlaybackSpeed.value != playbackParameters.speed
                    && settings.isPlaybackSpeedEnabled
                ) {
                    savedPlaybackSpeed.onNext(playbackParameters.speed)
                }
                if (savedPlaybackPitch.value != playbackParameters.pitch
                    && settings.isPlaybackPitchEnabled
                ) {
                    savedPlaybackPitch.onNext(playbackParameters.pitch)
                }
            }
        })

        exoPlayer.addAudioListener(object : AudioListener {
            override fun onAudioSessionId(audioSessionId: Int) {
                bassBoost = try {
                    BassBoost(0, audioSessionId).apply {
                        enabled = settings.isBassBoostEnabled
                        setStrength(settings.bassBoostStrength)
                    }
                } catch (e: RuntimeException) {
                    null
                }

                virtualizer = try {
                    Virtualizer(0, audioSessionId).apply {
                        enabled = settings.isVirtualizerEnabled
                        setStrength(settings.virtualizerStrength)
                    }
                } catch (e: RuntimeException) {
                    null
                }

                equalizer = try {
                    EqualizerEffectImpl(
                        audioSessionId,
                        enableOnStart = settings.isEqEnabled,
                        settings.savedBandLevels,
                    )
                } catch (e: RuntimeException) {
                    EqualizerEffectUnavailable()
                }
            }
        })
    }


    //    SPEED
    override fun observeIsPlaybackSpeedEnabled(): Observable<Boolean> = playbackSpeedEnabledUpdates

    override fun setPlaybackSpeedEnabled(enable: Boolean) {
        settings.isPlaybackSpeedEnabled = enable
        playbackSpeedEnabledUpdates.onNext(enable)

        val speed = if (enable) settings.playbackSpeed else DEFAULT_PLAYBACK_SPEED

        val playbackParameters = PlaybackParameters(speed, exoPlayer.playbackParameters.pitch)
        exoPlayer.setPlaybackParameters(playbackParameters)
    }

    override fun getSavedPlaybackSpeed(): Float = savedPlaybackSpeed.value!!

    override fun observePlaybackSpeed(): Observable<Float> = savedPlaybackSpeed

    override fun setSavedPlaybackSpeed(speed: Float) {
        settings.playbackSpeed = speed

        if (settings.isPlaybackSpeedEnabled) {
            val playbackParameters = PlaybackParameters(speed, exoPlayer.playbackParameters.pitch)
            exoPlayer.setPlaybackParameters(playbackParameters)
        }
    }


    //    PITCH
    override fun observeIsPlaybackPitchEnabled(): Observable<Boolean> = playbackPitchEnabledUpdates

    override fun setPlaybackPitchEnabled(enabled: Boolean) {
        settings.isPlaybackPitchEnabled = enabled
        playbackPitchEnabledUpdates.onNext(enabled)

        val pitch = if (enabled) settings.playbackPitch else DEFAULT_PLAYBACK_PITCH

        val playbackParameters = PlaybackParameters(exoPlayer.playbackParameters.speed, pitch)
        exoPlayer.setPlaybackParameters(playbackParameters)
    }

    override fun getSavedPlaybackPitch(): Float {
        return savedPlaybackPitch.value!!
    }

    override fun observePlaybackPitch(): Observable<Float> {
        return savedPlaybackPitch
    }

    override fun setSavedPlaybackPitch(pitch: Float) {
        settings.playbackPitch = pitch
        if (settings.isPlaybackPitchEnabled) {
            val playbackParameters = PlaybackParameters(
                exoPlayer.playbackParameters.speed, pitch
            )
            exoPlayer.setPlaybackParameters(playbackParameters)
        }
    }


    //    BASS BOOST
    override fun isBassBoostAvailable(): Boolean = bassBoost?.strengthSupported ?: false

    override fun isBassBoostEnabled(): Boolean = bassBoost?.enabled ?: false

    override fun setBassBoostEnable(enable: Boolean) {
        bassBoost?.enabled = enable
        settings.isBassBoostEnabled = enable
    }

    override fun observeBassBoostLevel(): Observable<Short> = savedBassBoostLevelUpdates

    override fun setBassBoostLevel(strength: Short) {
        if (isBassBoostAvailable) bassBoost!!.setStrength(strength)

        settings.bassBoostStrength = strength
        savedBassBoostLevelUpdates.onNext(strength)
    }


    //    VIRTUALIZER
    override fun isVirtualizerAvailable(): Boolean = virtualizer?.strengthSupported ?: false

    override fun isVirtualizerEnabled(): Boolean = virtualizer?.enabled ?: false

    override fun setVirtualizerEnable(enable: Boolean) {
        virtualizer?.enabled = enable
        settings.isVirtualizerEnabled = enable
    }

    override fun observeVirtualizerLevel(): Observable<Short> = savedVirtualizerLevelUpdates

    override fun setVirtualizerLevel(strength: Short) {
        if (isVirtualizerAvailable) virtualizer!!.setStrength(strength)

        settings.virtualizerStrength = strength
        savedVirtualizerLevelUpdates.onNext(strength)
    }


    //    EQ
    override fun isEqAvailable() = equalizer.isAvailable

    override fun setEqEnable(enable: Boolean) {
        equalizer.isEnabled = enable
        settings.isEqEnabled = enable
    }

    override fun getMinEqBandLevel() = equalizer.minBandLevel

    override fun getMaxEqBandLevel() = equalizer.maxBandLevel

    override fun observeEqEnabling() = equalizer.enablingUpdates

    override fun setBandLevel(bandId: Int, bandLevel: Int) {
        equalizer.setBandLevel(bandId, bandLevel)

        val bandLevels = equalizer.bands.map { it.currentLevel }
        settings.saveBandLevels(bandLevels)
    }

    override fun getBands(): List<EqBand> = equalizer.bands

    override fun usePreset(presetIndex: Int) {
        equalizer.usePreset(presetIndex)

        val bandLevels = equalizer.bands.map { it.currentLevel }
        settings.saveBandLevels(bandLevels)
    }

    override fun getPresets(): List<String> = equalizer.presets

}
